authmanager/proxy/proxy.go

```diff
@@ -0,0 +1,257 @@
+)
+
+// proxyConf holds the configuration of the local proxy
+type proxyConf struct {
+	backend  *url.URL
+	jwtToken string
+	listener net.Listener
+}
+
+// ProxyController proxies device API requests to Mender server
+type ProxyController struct {
+	conf   *proxyConf
+	client *http.Client
+	server *http.Server
+	done   chan struct{}
+
+	wsDialer           *websocket.Dialer
+	wsConnections      map[*wsConnection]bool
+	wsConnectionsMutex sync.Mutex
+}
+
+type wsConnection struct {
+	connClient            *websocket.Conn
+	connClientWriteMutex  sync.Mutex
+	connBackend           *websocket.Conn
+	connBackendWriteMutex sync.Mutex
+}
+
+func copyResponse(rw http.ResponseWriter, resp *http.Response) error {
+	copyHeader(rw.Header(), resp.Header)
+	rw.WriteHeader(resp.StatusCode)
+	defer resp.Body.Close()
+
+	_, err := io.Copy(rw, resp.Body)
+	return err
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+func (pc *ProxyController) DoHttpRequest(w http.ResponseWriter, r *http.Request) {
+	// Convert server request to client request and override r.URL
+	r.RequestURI = ""
+	r.Host = ""
+	r.URL.Scheme = pc.conf.backend.Scheme
+	r.URL.Host = pc.conf.backend.Host
+	log.Debugf("Request: %q %q %q %q %q", r.RequestURI, r.Host, r.URL.Scheme, r.URL.Host, r.URL.Path)
+
+	// Do the request
+	rsp, err := pc.client.Do(r)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	// Copy response to the user
+	_ = copyResponse(w, rsp)
+}
+
+func NewProxyController(client *http.Client, dialer *websocket.Dialer, menderUrl, menderJwtToken string) (*ProxyController, error) {
+	l, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create proxy")
+	}
+
+	pc := &ProxyController{
+		client:   client,
+		wsDialer: dialer,
+		conf: &proxyConf{
+			listener: l,
+		},
+		done:          make(chan struct{}, 1),
+		wsConnections: make(map[*wsConnection]bool),
+	}
+
+	if menderUrl != "" && menderJwtToken != "" {
+		err = pc.configure(menderUrl, menderJwtToken)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create proxy")
+		}
+	}
+
+	return pc, nil
+}
+
+func (pc *ProxyController) getPort() int {
+	return pc.conf.listener.Addr().(*net.TCPAddr).Port
+}
+
+// GetServerUrl returns the URL of the proxy
+func (pc *ProxyController) GetServerUrl() string {
+	return fmt.Sprintf("http://%s:%d", ProxyHost, pc.getPort())
+}
+
+// Start starts the local proxy server
+func (pc *ProxyController) Reconfigure(menderUrl, menderJwtToken string) error {
+	if pc.wsRunning() {
+		// Drop connections, force peers to re-connect
+		pc.CloseWsConnections()
+	}
+	return pc.configure(menderUrl, menderJwtToken)
+}
+
+func (pc *ProxyController) configure(menderUrl, menderJwtToken string) error {
+	u, err := url.Parse(menderUrl)
+	if err != nil {
+		return errors.Wrap(err, "failed to configure proxy")
+	}
+
+	pc.conf.backend = u
+	pc.conf.jwtToken = menderJwtToken
```

Comment:
There might be a race condition here too. These two values are changed here, but also used from the handlers in `start()`, which are running in a different Go routine.

Locking around these could also help prevent new connections from being established, as I pointed out in `Reconfigure()`.